### ISSUE TITLE
Document CGRect, etc. that have Swift methods

### ISSUE THREAD

**Issue by kballard** [SwiftDocOrg/swiftdoc.org#8]
There are various types from Foundation that gain extra methods in Swift. For example, `CGRect` has a bunch of methods/properties. Here's the definition as Swift sees it:

```swift
struct CGRect {
    var origin: CGPoint
    var size: CGSize
}

extension CGRect {
    static var zeroRect: CGRect { get }
    static var nullRect: CGRect { get }
    static var infiniteRect: CGRect { get }
    init()
    init(x: CGFloat, y: CGFloat, width: CGFloat, height: CGFloat)
    init(x: Double, y: Double, width: Double, height: Double)
    init(x: Int, y: Int, width: Int, height: Int)
    var width: CGFloat { get }
    var height: CGFloat { get }
    var minX: CGFloat { get }
    var midX: CGFloat { get }
    var maxX: CGFloat { get }
    var minY: CGFloat { get }
    var midY: CGFloat { get }
    var maxY: CGFloat { get }
    var isNull: Bool { get }
    var isEmpty: Bool { get }
    var isInfinite: Bool { get }
    var standardizedRect: CGRect { get }
    mutating func standardize()
    var integerRect: CGRect { get }
    mutating func integerize()
    func rectByInsetting(#dx: CGFloat, dy: CGFloat) -> CGRect
    mutating func inset(#dx: CGFloat, dy: CGFloat)
    func rectByOffsetting(#dx: CGFloat, dy: CGFloat) -> CGRect
    mutating func offset(#dx: CGFloat, dy: CGFloat)
    func rectByUnion(withRect: CGRect) -> CGRect
    mutating func union(withRect: CGRect)
    func rectByIntersecting(withRect: CGRect) -> CGRect
    mutating func intersect(withRect: CGRect)
    func rectsByDividing(atDistance: CGFloat, fromEdge: CGRectEdge) -> (slice: CGRect, remainder: CGRect)
    func contains(rect: CGRect) -> Bool
    func contains(point: CGPoint) -> Bool
    func intersects(rect: CGRect) -> Bool
}

extension CGRect : Equatable {
}

extension CGRect : Reflectable {

    /// Returns a mirror that reflects `self`.
    func getMirror() -> MirrorType
}
```

The other types defined in `CoreGraphics.CGGeometry` also have Swift additions, although far less than `CGRect` (mostly just an appropriate "zero" static and protocol conformance).

It would be extremely useful if SwiftDoc.org could document this stuff similar to how it documents Foundation extensions to Swift types (e.g. `String`).

**Comment by kballard** [comment]
Similarly, the `ObjectiveC` module has a function `autoreleasepool()` that should be documented. There may be other stuff there as well, I'm not sure (this function is notable for not being a translation of an Obj-C API, since Obj-C uses the builtin `@autoreleasepool` construct instead).

**Comment by natecook1000** [comment]
@kballard, thanks for opening this issue. I'm definitely straddling the fence on the question of SwiftDoc.org's scope, so I think I want to have a little more discussion before moving on this. My thoughts sort of for and against:

- I like that SwiftDoc.org is *nearly* only the Swift standard library. In the absence of actual documentation from Apple, I think it's important for there to be a resource to use both while coding and also to learn about what the Swift language offers on its own.

- I added the Foundation additions to `String` mainly because while `String` is a built-in type, the Foundation additions increase its capabilities a huge amount, and let's face it, if you're using Swift today you're probably also using Foundation. However, that `String` already exists in Swift feels like an important distinction.

- If we add `CGRect`, `CGPoint`, etc. to the types on SwiftDoc.org, where do we stop? If there are only a few additions in an extension, like for `NSArray`, is it confusing to only include those? Would we need to also include the rest of the `NSArray` declaration?

- At the same time, none of the `CGGeometry` changes are reflected anywhere else, and they're *really* useful. (I covered a chunk of them [in an update here](http://nshipster.com/cggeometry/).)

- There are also new types that are Swift-only but live in Foundation, like `NSFastGenerator` and `NSIndexSetGenerator`, as well as functions like `autoreleasepool()` and `NSLocalizedString()` that are different than the related concepts in Objective-C.

In short, I'm not sure where I stand. :smile: Let's talk a little more!

**Comment by uliwitness** [comment]
While I generally would prefer (especially for beginners) if SwiftDoc covered the whole frameworks, given the existence of third-party Swift implementations like Silver (from RemObjects), I would recommend clearly marking things that aren't part of the standard library. The SelfHTML tutorial did that nicely: They had an icon for each browser/browser version and showed that at the top of each entry. So you could immediately see whether a call is available to you.

**Comment by airspeedswift** [comment]
In a way, I feel better about including extensions to `CGRect` than I do about those `String` extensions.  In the case of `CGRect`, it is a struct, and the extensions are pure operations on that struct, with the same representation as in C.  The `String` extensions make me feel a little uneasy given the possibility they’re converting your `String` into an `NSString` and then doing operations on that.

**Comment by kballard** [comment]
My general feeling is that when doing iOS development, I should be able to see all of the documentation on stuff I can use by using the iOS docset and swiftdoc.org (or more accurately, the Dash docset generated from swiftdoc.org). Since the iOS docset doesn't document the Swift-specific extensions to CGGeometry types (or `autoreleasepool` and other things of that nature) I want swiftdoc.org to handle that.

I do see the utility in having documentation on just "pure Swift", but swiftdoc.org has already broken away from that by including Foundation extensions to `String`. That gives precedent to documenting other Swift-only things that come from the Foundation module (or one of the modules it includes, such as ObjectiveC).

In an ideal world, SwiftDoc would have a toggle that says "include Foundation things" that, when disabled, would give only pure Swift documentation (including removing the Foundation extensions on `String`). But I don't think that's useful enough to be worth the effort of implementing. Instead SwiftDoc should just be certain to clearly indicate which APIs come from Foundation.

**Comment by natecook1000** [comment]
@kballard I think you have me convinced! I'll leave this open while I get the changes built, in case anyone has any other feedback. If you want, feel free to submit a PR with the declarations you'd like to see in a new file in the [data/addenda directory](https://github.com/SwiftDocOrg/swiftdoc-parser/tree/master/data/addenda).
